File: apps/party-time-backend/src/main/java/com/partytime/api/error/ApiError.java
package com.partytime.api.error;

import com.fasterxml.jackson.annotation.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.http.HttpStatus;

import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Base Api Error Model
 */
@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class ApiError implements Serializable {

    private HttpStatus status;
    private LocalDateTime timestamp;
    private String message;

    @JsonIgnore
    private Throwable parent;

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private List<Map<String, Object>> additionalInformation;

    public static ApiError notFound(String message) {
        return notFound(message, null);
    }

    public static ApiError notFound(String message, Throwable parent) {
        return status(message, HttpStatus.NOT_FOUND, parent);
    }

    public static ApiError badRequest(String message) {
        return badRequest(message, null);
    }

    public static ApiError badRequest(String message, Throwable parent) {
        return status(message, HttpStatus.BAD_REQUEST, parent);
    }

    public static ApiError status(String message, HttpStatus status, Throwable parent) {
        return ApiError.builder()
            .status(status)
            .timestamp(LocalDateTime.now())
            .message(message)
            .parent(parent)
            .build();
    }

    /**
     * Builds a default unauthorized Error
     *
     * @return a default unauthorized Api Error
     */
    public static ApiError unauthorized() {
        return ApiError.builder()
            .status(HttpStatus.UNAUTHORIZED)
            .message("Die Anfrage konnte nicht bearbeitet werden, da keine gültigen Authentifizierungsdaten für die angeforderte Ressource vorliegen!")
            .timestamp(LocalDateTime.now())
            .build();
    }

    /**
     * Builds a default forbidden Error
     *
     * @return a default forbidden Api Error
     */
    public static ApiError forbidden() {
        return forbidden(null);
    }

    /**
     * Builds a default forbidden Error
     *
     * @param parent The original Exception
     * @return a default forbidden Api Error
     */
    public static ApiError forbidden(Throwable parent) {
        return ApiError.builder()
            .status(HttpStatus.FORBIDDEN)
            .message("Die Anfrage konnte nicht bearbeitet werden, da Sie für die angeforderte Resource keine ausreichenden Berechtigungen besitzen!")
            .timestamp(LocalDateTime.now())
            .parent(parent)
            .build();
    }

    public void throwNow() throws ApiErrorException {
        throw new ApiErrorException(this);
    }

    public void throwNow(Throwable throwable) throws ApiErrorException {
        throw new ApiErrorException(this, throwable);
    }

    public ApiError withAdditionalInformation(Map<String, Object> additionalInformation) {
        if (this.additionalInformation == null) {
            this.additionalInformation = new ArrayList<>();
        }
        this.additionalInformation.add(additionalInformation);
        return this;
    }

    public ApiErrorException asException() {
        return new ApiErrorException(this);
    }


}

File: apps/party-time-backend-kotlin/src/main/kotlin/com/partytime/api/error/ApiError.kt
package com.partytime.api.error

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import org.springframework.http.HttpStatus
import java.io.Serializable
import java.time.ZoneOffset
import java.time.ZonedDateTime


class ApiError private constructor(
    val status: HttpStatus,
    val message: String,
    @JsonIgnore
    val parent: Throwable? = null,
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    val additionalInformation: List<Map<String, Any>>? = null
) : Serializable {
    companion object {

        fun notFound(
            message: String,
            parent: Throwable? = null,
        ) = ApiError(
            HttpStatus.NOT_FOUND,
            message,
            parent
        )

        fun badRequest(message: String, parent: Throwable? = null) = ApiError(
            HttpStatus.BAD_REQUEST,
            message,
            parent
        )

        fun unauthorized(message: String? = null) = ApiError(
            HttpStatus.UNAUTHORIZED,
            message
                ?: "Die Anfrage konnte nicht bearbeitet werden, da keine gültigen Authentifizierungsdaten für die angeforderte Ressource vorliegen!",
        )

        fun forbidden(parent: Throwable? = null) = ApiError(
            HttpStatus.FORBIDDEN,
            "Die Anfrage konnte nicht bearbeitet werden, da Sie für die angeforderte Resource keine ausreichenden Berechtigungen besitzen!",
            parent
        )

        fun conflict(message: String, parent: Throwable, additionalInformation: List<Map<String, Any>>) = ApiError(
            HttpStatus.CONFLICT,
            message,
            parent,
            additionalInformation
        )

        fun arbitraryCode(status: HttpStatus, message: String, parent: Throwable) = ApiError(
            status,
            message,
            parent
        )

    }

    val timeStamp: ZonedDateTime = ZonedDateTime.now(ZoneOffset.UTC)
}
